### PR TITLE
[KYUUBI #5786] Disable spark script transformation

### LINKED ISSUE #5786: [TASK][EASY] Disable spark ScriptTransformation
### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before asking

- [X] I have searched in the [issues](https://github.com/apache/kyuubi/issues?q=is%3Aissue) and found no similar issues.

### What would you like to be improved?

Spark ScriptTransformation can use scripts to operate HDFS or other storage directories. The permissions cannot be controlled in Authz, which is more dangerous. So need to disabled.

### How should we improve?

Add check rule.

### Are you willing to submit PR?

- [X] Yes. I would be willing to submit a PR with guidance from the Kyuubi community to improve.
- [ ] No. I cannot submit a PR at this time.

### PR DESCRIPTION
**Pull request by zml1206**
# :mag: Description
## Issue References 🔗

This pull request fixes #5786.

## Describe Your Solution 🔧

Add spark check rule.

## Types of changes :bookmark:

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

#### Related Unit Tests
org.apache.kyuubi.plugin.spark.authz.rule.AuthzUnsupportedOperationsCheckSuite.test("disable script transformation")

---

# Checklists
## 📝 Author Self Checklist

- [x] My code follows the [style guidelines](https://kyuubi.readthedocs.io/en/master/contributing/code/style.html) of this project
- [x] I have performed a self-review
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] This patch was not authored or co-authored using [Generative Tooling](https://www.apache.org/legal/generative-tooling.html)

## 📝 Committer Pre-Merge Checklist

- [ ] Pull request title is okay.
- [ ] No license issues.
- [ ] Milestone correctly set?
- [ ] Test coverage is ok
- [ ] Assignees are selected.
- [ ] Minimum number of approvals
- [ ] No changes are requested

**Be nice. Be informative.**

### CODE REVIEW

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/rule/AuthzUnsupportedOperationsCheck.scala [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ */
+
+package org.apache.kyuubi.plugin.spark.authz.rule
+
+import org.apache.spark.sql.catalyst.plans.logical.{LogicalPlan, ScriptTransformation}
+
+import org.apache.kyuubi.plugin.spark.authz.AccessControlException
+
+class AuthzUnsupportedOperationsCheck extends (LogicalPlan => Unit) {
+  override def apply(plan: LogicalPlan): Unit = plan foreach {
```
Comment (cxzl25) [inline extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/rule/AuthzUnsupportedOperationsCheck.scala:25]: We can add a configuration to decide whether to enable it.

Comment (zml1206) [reply extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/rule/AuthzUnsupportedOperationsCheck.scala:25]: When authz is enabled, this is considered a security vulnerability, so should it be forcibly disabled?

Comment (zml1206) [reply extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/rule/AuthzUnsupportedOperationsCheck.scala:25]: How about add a configuration default disabled it?

**Comment by zml1206** [comment]
> in spark extensions plugin

Why in spark extensions plugin? @wForget

**Comment by wForget** [comment]
> Why in spark extensions plugin? @wForget

As far as I understand, the authz plugin implements access control by integrating ranger. This PR rejects ScriptTransformation command based on configuration, more like the watchdog in the spark extensions plugin.

**Comment by ulysses-you** [comment]
If the code is actually not related to auth, we can move this new rule to extension module and revisit the config name. It can help people to ban script transform without auth module.

**Comment by zml1206** [comment]
Updated, @wForget @ulysses-you

**Review by ulysses-you — APPROVED** [review]
lgtm, thank you @zml1206

**Review by wForget — APPROVED** [review]
LGTM, Thanks.

File: extensions/spark/kyuubi-extension-spark-3-4/src/main/scala/org/apache/kyuubi/sql/watchdog/KyuubiUnsupportedOperationsCheck.scala [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+ */
+
+package org.apache.kyuubi.sql.watchdog
+
+import org.apache.spark.sql.catalyst.SQLConfHelper
+import org.apache.spark.sql.catalyst.plans.logical.{LogicalPlan, ScriptTransformation}
+
+import org.apache.kyuubi.sql.{KyuubiSQLConf, KyuubiSQLExtensionException}
+
+class KyuubiUnsupportedOperationsCheck extends (LogicalPlan => Unit) with SQLConfHelper {
```
Comment (pan3793) [inline extensions/spark/kyuubi-extension-spark-3-4/src/main/scala/org/apache/kyuubi/sql/watchdog/KyuubiUnsupportedOperationsCheck.scala:25]: possible to define it as `object`?

**Comment by wForget** [comment]
thanks, merged to master.

File: extensions/spark/kyuubi-extension-spark-3-4/src/main/scala/org/apache/kyuubi/sql/watchdog/KyuubiUnsupportedOperationsCheck.scala (hunk 2 above)
Comment (wForget) [reply extensions/spark/kyuubi-extension-spark-3-4/src/main/scala/org/apache/kyuubi/sql/watchdog/KyuubiUnsupportedOperationsCheck.scala:25]: Sorry this pr has been merged, @zml1206 do you have time to submit a follow-up pr to improve it?

Comment (zml1206) [reply extensions/spark/kyuubi-extension-spark-3-4/src/main/scala/org/apache/kyuubi/sql/watchdog/KyuubiUnsupportedOperationsCheck.scala:25]: Okay, I see there are other classes that also need to be changed. I will modify them together later. @pan3793 @wForget

**Comment by zml1206** [comment]
Thanks all, @cxzl25 @wForget @ulysses-you @pan3793

**Pull request closed without merging**